src/puya/awst_build/utils.py

```diff
@@ -339,3 +345,199 @@ def get_arg_mapping(
                 raise CodeError("Too many positional arguments", location) from ex
         arg_mapping[arg_name] = arg
     return arg_mapping
+
+
+ALLOWABLE_OCA = [oca.name for oca in OnCompletionAction if oca != OnCompletionAction.ClearState]
+
+
+def get_func_types(
+    context: ASTConversionModuleContext, func_def: mypy.nodes.FuncDef, location: SourceLocation
+) -> dict[str, wtypes.WType]:
+    start_idx = 0
+    if func_def.arguments:
+        first_arg_var = func_def.arguments[0].variable
+        if first_arg_var.is_self or first_arg_var.is_cls:
+            start_idx = 1
+    in_var_names = [arg.variable.name for arg in func_def.arguments[start_idx:]]
+    if "output" in in_var_names:
+        # https://github.com/algorandfoundation/ARCs/blob/main/assets/arc-0032/application.schema.json
+        raise CodeError(
+            "For compatibility with ARC-32, ARC-4 methods cannot have an argument named output",
+            location,
+        )
+    names = [*in_var_names, "output"]
+    match func_def.type:
+        case mypy.types.CallableType(arg_types=arg_types, ret_type=ret_type):
+            types = arg_types[start_idx:] + [ret_type]
+            wtypes_ = (context.type_to_wtype(t, source_location=location) for t in types)
+            return dict(zip(names, wtypes_, strict=True))
+    raise InternalError("Unexpected FuncDef type")
+
+
+def get_arc4_method_config(
```

Comment:
Should we split utils into a module and put these in separate files? It's getting a bit big

Reply:
Done